share/man/man1/setoraenv.1
.\" vim:ts=4:sw=4:syntax=nroff
.fp 1 R
.fp 2 I
.fp 3 B
.fp 4 BI
.fp 5 CO
.fp 6 CI
.fp 7 CB
.nr X
.TH setoraenv 1 "9 Jun 2026" ""
.SH NAME
setoraenv \- set environment for database access
.SH SYNOPSIS
\f3. /etc/profile.d/setoraenv.sh\f1
.sp
\f3setoraenv\f1 [\f3-l\f1] [\f3-i\f1[\f3-u\f1]|-s \f2oracle_sid\f1 [\f3-e\f1[\f3-o\f1][\f3-w\f1[\f3-x\f1]]]
.SH DESCRIPTION
.IX "setoraenv"
.sp
.P
\f3setoraenv\f1 uses \f3oraenv\f1 to set the current shell environment with approriate values for a valid \f3ORACLE_HOME\f1.
.P
.SH OPTIONS
.TP
\f3-l\f1
Consider the ORACLE_SID local. That is a valid ORACLE_HOME must be present for
the BEQ adapater to be applicable.
.TP
\f3-i\f1
Interactively select an instance from a menu built from the contents of the oratab.
.IP
For example:
.DS
.sp
.nf
\f1$ \f3setoraenv -i\f1

  1  FOO        /opt/oracle/product/19c/dbhome
  2  agent      /opt/oracle/product/13c/agent/agent_13.5.0.0.0
  3  aiorcl     /opt/oracle/product/26ai/dbhome
  4  oms        /opt/oracle/product/24ai/mw/oms_home
  5  orcl       /opt/oracle/product/19c/dbhome

Choose an instance [?,??,q]:
.fi
.DE
.TP
\f3-u\f1
Do not sort the interactive menu generated with \f3-i\f1.
.TP
\f3-s \f2oracle_sid\f1
Supply an ORACLE_SID for \f3oraenv\f1 to set the environment.
.TP
\f3-e\f1
Set additional environment variables read from a file, \f6envfile\f1, to be found in a directory defined by the environment variable \f5ENVDIR\f1 (or a default location).
.IP
This file is typically named \f2oracle_sid\f3.env\f1, however, using \f2oracle_sid\f1 is not mandatory since the key:
.DS
.nf
.sp
 \f5ORACLE_SID=\f1
.fi
.DE
.IP 
is used to locate the file. There must be only one such \f5ORACLE_SID=\f1 entry in all files located in \f5ENVDIR\f1.
.IP
This file has the format:
.DS
.sp
.nf
 \f2VAR1\f1=\f2value1\f1
 \f2VAR2\f1=\f2value2\f1
 \f2VAR3\f1=\f2value3\f1
 ...
.fi
.DE
.IP
Lines starting with '#' are comments. Blank lines will also be ignored.
.TP
\f3-o\f1
When setting the environment any settings in the \f6envfile\f1 will override those
made by \f3oraenv\f1. This flag ensures that the \f3oraenv\f1 settings take
precedence.
.TP
\f3-w\f1
Write a copy of the environment settings from the original \f6envfile\f1 (with
possible modifications made by \f3oraenv\f1) to
the home directory of the current user.
.TP
\f3-x\f1
In the script created when using \f3-w\f1 append \f5export \f2var\f1 to each
assignment. Ordinarily it might be expected to use the shell to mark the
assigments for export with \f3set -a\f1 before sourcing the file.
.SH FILES
.TP 30
\f5/etc/oratab\f1
oratab file on Linux systems
.TP 30
\f5/etc/opt/oracle/env\f1
Default location for environment files.
.TP 30
\f5/usr/local/bin/oraenv\f1
Default location for \f3oraenv\f1 program.
.TP 30
\f5/usr/local/bin/dbhome\f1
Default location for \f3dbhome\f1 program.
.SH ENVIRONMENT
.TP 10 
ENVDIR
Location for .env files. Default is @PKGDATADIR@.
.SH EXAMPLES
Set the environment for the named database and write an \f5envfile\f1:
.nf

\f5$ \f3setoraenv -s FOO -e -w\f1\f5
The Oracle base has been changed from /u02/app/oracle to /var/opt/oracle/dbhome
writing /export/home/oracle/foo.env...
PATH=/usr/bin:/usr/local/bin export PATH
ORACLE_UNQNAME=FOO
ORACLE_SID=FOO
ORACLE_HOSTNAME=oel.ipa.home
LD_LIBRARY_PATH=/opt/oracle/product/19c/dbhome/lib
ORACLE_BASE=/u02/app/oracle
OH=/opt/oracle/product/19c/dbhome
ORACLE_HOME=/opt/oracle/product/19c/dbhome
TNS_ADMIN=/etc/opt/oracle/network/admin\f1
.fi
.P
The same, except with settings made by \f3oraenv\f1 taking precedence:
.nf
\f5$ \f3setoraenv -s FOO -e -w -o\f1\f5
The Oracle base has been changed from /u02/app/oracle to /var/opt/oracle/dbhome
writing /export/home/oracle/foo.env...
PATH=/usr/bin:/usr/local/bin:/opt/oracle/product/19c/dbhome/bin
ORACLE_UNQNAME=FOO
ORACLE_SID=FOO
ORACLE_HOSTNAME=oel.ipa.home
LD_LIBRARY_PATH=/opt/oracle/product/19c/dbhome/lib
ORACLE_BASE=/var/opt/oracle/dbhome
OH=/opt/oracle/product/19c/dbhome
ORACLE_HOME=/opt/oracle/product/19c/dbhome
TNS_ADMIN=/opt/oracle/product/19c/dbhome/network/admin
.fi
.SH AUTHOR
Simon Anthony(simon.a.anthony@oracle.com)
.SH SEE ALSO
.BR bash (1)
.SH NOTES
.P
The use of the \f3-i\f1 options requires the System V valtools to be
installed.
.P
In the examples given above the following variables need not be set as they
are set by \f3oraenv\f1:
.nf
.fi
.TP 16
LD_LIBRARY_PATH
Unless additional shared objects for other programs need to be found
.TP 16
ORACLE_BASE
Set by \f3oraenv\f1
.TP 16
ORACLE_HOME
Set by \f3oraenv\f1
.TP 16
OH
This is the same as \f5ORACLE_HOME\f1 and could be set in the "custom" section
of \f3oratab\f1 if required - see the note below.
.P
In addition, there are other cases where, for example, some users should have
different settings for a variable.
.P
This is demonstrated in the above example:
.TP 16
TNS_ADMIN
Is set by \f3oraenv\f1 in its "custom" section:
.DS
.nf
.sp
#
# Install any "custom" code here
#
export TNS_ADMIN=$ORACLE_HOME/network/admin
.fi
.DE
.IP 
The use of \f3-o\f1 determines which setting of TNS_ADMIN is set in the
environment or written to the local file.
.IP
The directory \f5/usr/local/bin\f1 containing \f3oraenv\f1 and \f3dbhome\f1
must be in the \f5PATH\f1 of the current shell as \f3setoraenv\f1 is a shell
function and does not set its own environment.


etc/profile.d/setoraenv.sh
# vim:syntax=sh:sw=4:ts=4:
################################################################################
# setoraenv: interactive invocation of oraenv
#
################################################################################
# Copyright (C) 2026
#
# This program is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation, either version 3 of the License or, (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
# Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program. If not see <http://www.gnu.org/licenses/>>
#

readoraenv() {
	typeset -r prog="readoraenv"
	local var= val=

	[ -r "$1" ] || return
	while IFS="=" read -r var val
	do
		set -a
		eval $var=$val
		set +x
	done < <(sed -n '/^[A-Z_]\{1,\}=/ {
		s;\([^=]*\)="\([^"]*\)";\1="";
		s%\([^=]*\)=\([^;]*\)[\;&|].*%\1=\2%
		p; }' $1)
}

writeoraenv() {
	typeset -r prog="writeoraenv"
	local eflg= xflg= opt= errflg= OPTIND=1
	local var= val=

	while getopts "ex" opt 2>&-
	do
		case $opt in
		e)	eflg=y ;;
		x)	xflg=y ;;
		\?)	errflg=y
		esac
	done
	shift $(( OPTIND - 1 )) 

	[ $# -eq 1 ] || errflg=y
	[ $errflg ] && { echo "usage: $prog [-ex] <file>" >&2; return 2; }

	local envfile=$HOME/`basename $1` 

	[ -r "$1" ] || return 1
	> $envfile || return

	echo writing $envfile ...
	while IFS="=" read -r var val
	do
		if [ $eflg ]
		then
			eval echo $var=\$$var ${xflg:+export $var} | tee -a $envfile
		else
			eval echo $var=$val ${xflg:+export $var} | tee -a $envfile
		fi
	done < <(sed -n '/^[A-Z_]\{1,\}=/ {
		s;\([^=]*\)="\([^"]*\)";\1="";
		s%\([^=]*\)=\([^;]*\)[\;&|].*%\1=\2%
		p; }' $1)
}

setoraenv() {
	typeset -r prog="setoraenv"
	local iflg= uflg= sflg= lflg= eflg= oflg= wflg= xflg= opt= errflg= OPTIND=1
	local oracle_sid=
	local envdir="${ENVDIR:-@PKGDATADIR@}"
	local envfile=
	local suffix="env"
	
	while getopts "ius:leowx" opt 2>&-
	do
		case $opt in
		i)	iflg=y
			[ $sflg ] && errflg=y
			;;
		u)	uflg=y
			;;
		s)	sflg=y	
			oracle_sid="$OPTARG"
			[ $iflg ] && errflg=y
			;;
		l)	lflg=y	# consider oracle_sid local
			;;
		e)	eflg=y	# source an envfile
			envfile=y
			;;
		o)	oflg=y	# oraenv settings will take preference over envfile
			;;
		w)	wflg=y	# write resulting envfile 
			;;
		x)	xflg=y	# add export to script
			;;
		\?)	errflg=y
		esac
	done
	shift $(( OPTIND - 1 )) 

	[ $# -eq 0 ] || errflg=y
	[ -n "$uflg" -a -z "$iflg" ] && errflg=y
	[ -n "$eflg" -a -z "$sflg" ] && errflg=y # -e requires -s
	[ -n "$wflg" -a -z "$eflg" ] && errflg=y # -w requires -e 
	[ -n "$oflg" -a -z "$eflg" ] && errflg=y # -o requires -e 
	[ -n "$xflg" -a -z "$wflg" ] && errflg=y # -x requires -w 

	[ $errflg ] && { echo "usage: $prog [-l] [-i[-u]|-s <sid> [-e[-o][-w[-x]]]" >&2; return 2; }

	local oratab=`ls /etc/oratab /var/opt/oracle/oratab 2>&-`

    [ -x /usr/local/bin/oraenv ] || { echo "$prog: no oraenv in local bin" >&2; return 1; }
	[ -n "$oratab" -a -r "$oratab" ] || { echo "$prog: no oratab" >&2; return 1; }

	if [ $iflg ]
	then
		[ -x /usr/local/bin/ckitem ] || { echo "$prog: valtools package not installed" >&2; return 1; }
		local file=`mktemp`

		awk -F: '
			/^[ ]*#/ { next; }
			$1 ~ /[A-Za-z0-9]+/ { printf("%s\t%s\n", $1, $2); }' /etc/oratab > $file

		oracle_sid=`ckitem ${uflg:+-u} -o -p "Choose an instance" -f $file` || return
		echo oracle_sid is $oracle_sid
		rm -f $file
	elif [ $sflg ]
	then
		oracle_sid=`awk -F: '$1 == "'$oracle_sid'" { printf("%s", $1); quit; }' $oratab`
		if [ "$envfile" ]
		then
			envfile=`grep -s -l ORACLE_SID=''$oracle_sid'[\ ;$]' $envdir/*.$suffix`
			case $? in
			1) echo "no environment file containing ORACLE_SID='$oracle_sid'" >&2; wflg= ;;
			2) echo "no environment files found" >&2; wflg= ;;
			esac
			if [ "${envfile#*.$suffix}" ] # too many matches
			then
				echo "ORACLE_SID '$oracle_sid' found in too many files: $envfile" >&2
				return 1
			fi
		fi
	else # fail
		return 1
	fi

	if [ -n "$oracle_sid" ]
	then 
		[ -n "$eflg" -a -n "$oflg" ] && readoraenv $envfile
		export ORACLE_SID="$oracle_sid" ORAENV_ASK=NO
		. oraenv 
		[ -n "$eflg" -a -z "$oflg" ] && readoraenv $envfile
	else
		echo "$prog: specified SID not in $oratab"
		return 1
	fi
	[ $wflg ] && writeoraenv ${oflg:+-e} ${xflg:+-x} $envfile
	if [ ! "$lflg" ]
	then
		# ORACLE_SID is only for local instances with BEQ connection
		[ -f $ORACLE_HOME/dbs/hc_$ORACLE_SID.dat ] || unset ORACLE_SID # is instance local?
	fi
	[ $oracle_sid ] || return 1
}

typeset -fx setoraenv readoraenv writeoraenv
